File: product-brain/src/product_brain/backfill/git_log.py
from __future__ import annotations

import re
import subprocess
from collections import defaultdict
from datetime import datetime
from pathlib import Path
from typing import Optional

from ..models import Commit, FileChange


_RECORD_SEP = "\x1e\x1e\x1e"
_FIELD_SEP = "\x1f"


def _git_log_cmd(repo: Path, workflow: str, since: Optional[str] = None) -> list[str]:
    cmd = ["git", "-C", str(repo), "log", "--all"]
    if workflow == "merge" or workflow == "rebase":
        cmd.append("--no-merges")
    cmd += [
        f"--pretty=format:{_RECORD_SEP}%H{_FIELD_SEP}%an{_FIELD_SEP}%ae{_FIELD_SEP}%aI{_FIELD_SEP}%P{_FIELD_SEP}%s{_FIELD_SEP}%b",
        "--name-status",
        "--find-renames=50%",
    ]
    if since:
        cmd.append(f"{since}..HEAD")
    return cmd


def parse_git_log(repo: Path, ticket_regex: str, workflow: str = "squash", since: Optional[str] = None) -> list[Commit]:
    res = subprocess.run(_git_log_cmd(repo, workflow, since), capture_output=True, text=True, errors="replace")
    if res.returncode != 0:
        raise RuntimeError(f"git log failed: {res.stderr}")
    pat = re.compile(ticket_regex)

    commits: list[Commit] = []
    chunks = res.stdout.split(_RECORD_SEP)
    for chunk in chunks:
        chunk = chunk.strip("\n")
        if not chunk:
            continue
        first_line, _, rest = chunk.partition("\n")
        fields = first_line.split(_FIELD_SEP)
        if len(fields) < 7:
            continue
        sha, author, email, date_iso, parents_str, subject, body = fields[:7]
        date = datetime.fromisoformat(date_iso)
        parents = parents_str.split() if parents_str else []
        files: list[FileChange] = []
        for line in rest.strip("\n").splitlines():
            parts = line.split("\t")
            if len(parts) < 2:
                continue
            code = parts[0]
            if code == "A":
                files.append(FileChange(path=parts[1], change="added"))
            elif code == "M":
                files.append(FileChange(path=parts[1], change="modified"))
            elif code == "D":
                files.append(FileChange(path=parts[1], change="deleted"))
            elif code.startswith("R") and len(parts) >= 3:
                files.append(FileChange(path=parts[2], change="renamed"))
        text = subject + "\n" + (body or "")
        tickets = sorted(set(pat.findall(text)))
        commits.append(Commit(
            sha=sha, author=author, author_email=email, date=date,
            subject=subject, body=body, parents=parents, files=files, tickets=tickets,
        ))
    return commits


def group_by_ticket(commits: list[Commit]) -> dict[str, list[Commit]]:
    out: dict[str, list[Commit]] = defaultdict(list)
    for c in commits:
        for t in c.tickets:
            out[t].append(c)
    for t in out:
        out[t].sort(key=lambda c: c.date)
    return out


def diff_stat(repo: Path, sha: str) -> dict[str, tuple[int, int]]:
    """Return {path: (added, removed)} for one commit. Used to fill loc fields."""
    res = subprocess.run(
        ["git", "-C", str(repo), "show", "--numstat", "--pretty=", sha],
        capture_output=True, text=True, errors="replace",
    )
    out: dict[str, tuple[int, int]] = {}
    for line in res.stdout.splitlines():
        parts = line.split("\t")
        if len(parts) >= 3:
            try:
                added = int(parts[0]) if parts[0].isdigit() else 0
                removed = int(parts[1]) if parts[1].isdigit() else 0
                out[parts[2]] = (added, removed)
            except ValueError:
                continue
    return out

File: product-brain/src/backfill/git-log.ts
import { execFileSync } from "node:child_process";
import type { Commit, FileChange } from "../models.js";

const RECORD_SEP = "\x1e\x1e\x1e";
const FIELD_SEP = "\x1f";

function gitArgs(workflow: string, since?: string): string[] {
  const args = ["log", "--all"];
  if (workflow === "merge" || workflow === "rebase") args.push("--no-merges");
  args.push(
    `--pretty=format:${RECORD_SEP}%H${FIELD_SEP}%an${FIELD_SEP}%ae${FIELD_SEP}%aI${FIELD_SEP}%P${FIELD_SEP}%s${FIELD_SEP}%b`,
    "--name-status",
    "--find-renames=50%",
  );
  if (since) args.push(`${since}..HEAD`);
  return args;
}

export function parseGitLog(
  repo: string,
  ticketRegex: string,
  workflow: "squash" | "merge" | "rebase" = "squash",
  since?: string,
): Commit[] {
  let stdout: string;
  try {
    stdout = execFileSync("git", ["-C", repo, ...gitArgs(workflow, since)], {
      encoding: "utf8",
      maxBuffer: 256 * 1024 * 1024,
    });
  } catch (e) {
    throw new Error(`git log failed: ${(e as Error).message}`);
  }
  const pat = new RegExp(ticketRegex, "g");
  const commits: Commit[] = [];
  for (const chunk of stdout.split(RECORD_SEP)) {
    const trimmed = chunk.replace(/^\n+|\n+$/g, "");
    if (!trimmed) continue;
    const nl = trimmed.indexOf("\n");
    const firstLine = nl === -1 ? trimmed : trimmed.slice(0, nl);
    const rest = nl === -1 ? "" : trimmed.slice(nl + 1);
    const fields = firstLine.split(FIELD_SEP);
    if (fields.length < 7) continue;
    const [sha, author, email, dateIso, parentsStr, subject, body] = fields as [
      string, string, string, string, string, string, string,
    ];
    const date = new Date(dateIso);
    const parents = parentsStr ? parentsStr.split(" ").filter(Boolean) : [];
    const files: FileChange[] = [];
    for (const line of rest.replace(/^\n+|\n+$/g, "").split("\n")) {
      const parts = line.split("\t");
      if (parts.length < 2) continue;
      const code = parts[0] ?? "";
      if (code === "A") files.push({ path: parts[1] ?? "", change: "added" });
      else if (code === "M") files.push({ path: parts[1] ?? "", change: "modified" });
      else if (code === "D") files.push({ path: parts[1] ?? "", change: "deleted" });
      else if (code.startsWith("R") && parts.length >= 3) files.push({ path: parts[2] ?? "", change: "renamed" });
    }
    const text = `${subject}\n${body || ""}`;
    const matches = text.match(pat) ?? [];
    const tickets = [...new Set(matches)].sort();
    commits.push({
      sha,
      author,
      authorEmail: email,
      date,
      subject,
      body,
      parents,
      files,
      tickets,
    });
  }
  return commits;
}

export function groupByTicket(commits: Commit[]): Map<string, Commit[]> {
  const out = new Map<string, Commit[]>();
  for (const c of commits) {
    for (const t of c.tickets) {
      const arr = out.get(t) ?? [];
      arr.push(c);
      out.set(t, arr);
    }
  }
  for (const [, arr] of out) arr.sort((a, b) => a.date.getTime() - b.date.getTime());
  return out;
}

export function diffStat(repo: string, sha: string): Map<string, [number, number]> {
  let stdout: string;
  try {
    stdout = execFileSync("git", ["-C", repo, "show", "--numstat", "--pretty=", sha], {
      encoding: "utf8",
      maxBuffer: 64 * 1024 * 1024,
    });
  } catch {
    return new Map();
  }
  const out = new Map<string, [number, number]>();
  for (const line of stdout.split("\n")) {
    const parts = line.split("\t");
    if (parts.length < 3) continue;
    const added = /^\d+$/.test(parts[0] ?? "") ? Number(parts[0]) : 0;
    const removed = /^\d+$/.test(parts[1] ?? "") ? Number(parts[1]) : 0;
    out.set(parts[2] ?? "", [added, removed]);
  }
  return out;
}
